core/src/main/scala/com/stripe/dagon/Dag.scala

```diff
@@ -76,7 +76,17 @@ sealed abstract class Dag[N[_]] { self =>
   private def gc: Dag[N] = {
     val keepers = reachableIds
     if (idToExp.forallKeys(keepers)) this
-    else copy(id2Exp = idToExp.filterKeys(keepers))
+    else {
+      val id2Exp = idToExp.filterKeys(keepers)
+      // if we have GC'd an id it is because
+      // it did not escape. I.e. it is not a root,
+      // nor reachable from the root, so we can
+      // possibly lower the nextId, which is a bit
+      // nice because (unfortunately) Ids wrap Int
+      // not Long, so overflow is possible
+      val nextId = id2Exp.keySet.iterator.map(_.id).max
```

Comment:
Might be nice to add something to `HMap` to make this more efficient, but we can do that later.